### DISCUSSION TITLE
`DotLottieFile` init with Data

### DISCUSSION THREAD

**Discussion by kirti-swiggy** [airbnb/lottie-ios#2087] [Q&A] [answered]
For my use case, I need to download a dotLottie file from the remote and store it in a persistent cache so that I can load the lottie from the cache for the next launch(es). I am able to download the data (`Foundation.Data`) from the URL response and save it to the cache. But I couldn't find a way to create `DotLottieFile` instance using the data. There is an init of `DotLottieFile` that takes Data as a parameter but unfortunately, that's an internal function.

Please suggest if there's another way to do this. Thanks in advance!

**Comment by calda** [comment, marked as answer]
If you're interested in this functionality you could submit a pull request adding a helper for this to [DotLottieFileHelpers.swift](https://github.com/airbnb/lottie-ios/blob/master/Sources/Public/DotLottie/DotLottieFileHelpers.swift). It seems reasonable to have a helper for creating a `DotLottieFile` from a `Foundation.Data`, considering we also have a `Data` helper for `LottieAnimation` itself [here](https://github.com/airbnb/lottie-ios/blob/master/Sources/Public/Animation/LottieAnimationHelpers.swift#L152).

**Reply by kirti-swiggy** [reply]
Thanks much @calda for the quick reply! I'll raise the PR.

**Comment by nsarno-lucid** [comment]
I have a similar use-case where we want to init a DotLottieFile with Data, except it needs to be done synchronously.

This is possible using `LottieAnimation`...
```swift
case .json(let dict):
     if let animation = try? LottieAnimation(dictionary: dict) {
          LottieView(animation: animation)
     }
```

But the only init for `DotLottieFile` seems to be asynchronous...
```swift
case .data(data: let data):
     LottieView({
          try await .dotLottieFile(.loadedFrom(data: data, filename: "abc"))
     })
```

Is it possible to add a synchronous initializer for `DotLottieFile`? Something like this would be ideal...

```swift
case .data(data: let data):
     if let dotLottie = try? DotLottieFile(data: data) {
          LottieView(dotLottie: dotLottie)
     }
```

Please let me know if there's already a way to do this! Thanks.

**Reply by calda** [reply]
All DotLottie initializers are asynchronous because it involves interacting with the file system. A DotLottie files is really a zip archive, which has to be extracted to disk before we can load the individual files. There's a `DotLottieFile.SynchronouslyBlockingCurrentThread` namespace that loads a DotLottie file synchronously by blocking the thread while waiting for the file operations. You could use that instead if it suits your needs.

**Reply by nsarno-lucid** [reply]
Thanks for the quick response @calda! I will try to make this work.
